dagutils/traversal.go

```diff
@@ -0,0 +1,507 @@
+package dagutils
+
+import (
+	"context"
+	"errors"
+
+	cid "gx/ipfs/QmYVNvtQkeZ6AKSwDrjQTs432QtL6umrrK41EBq3cu7iSP/go-cid"
+	ipld "gx/ipfs/QmZtNq8dArGfnpCZfx2pUNY7UcjGhVp5qqwQ4hH6mpTMRQ/go-ipld-format"
+)
+
+// Traversal provides methods to move through a DAG of IPLD nodes
+// using an iterative algorithm and a handler for the user to perform
+// some logic on each visited node. This enables the implementation of
+// operations like reading and seeking inside of a UnixFS file DAG. In
+// contrast with a recursive solution, the implicit path in the call
+// stack is made explicit through the `path` variable.
+//
+// It exposes simple traversal operations like `Iterate` and  `Search`
+// that are implemented through a series of simple move methods (`down`,
+// `up`, `Right`) that modify its `path`. The `down` move method is the
+// analogous of the recursive call and the one in charge of visiting
+// the node (through the `VisitHandler`) and performing some user-defined
+// logic.
```

Comment:
`Right()` being the only one of these that is public feels a bit odd to me, but I get why it’s here. See my summary comment for some alternative thoughts.

Reply:
Yes, agree. Maybe some other public method like `SkipNode` should be exported (that would internally call an un-exported `right`), but I feel that the graphic "turn right" movement is something that the user could grasp rather easily and it could relate more to that than a `Skip` function. Not really sure though, I usually lean towards abstracting but this felt like a natural "take look under the hood, you see? there's not much complex logic down there".